Add subnet validations for DoS resistance (#1615)
ethereum/eth2.0-specs: specs/phase0/p2p-interface.md

- `beacon_aggregate_and_proof` - This topic is used to propagate aggregated attestations (as `SignedAggregateAndProof`s) to subscribing nodes (typically validators) to be included in future blocks. The following validations MUST pass before forwarding the `signed_aggregate_and_proof` on the network. (We define the following for convenience -- `aggregate_and_proof = signed_aggregate_and_proof.message` and `aggregate = aggregate_and_proof.aggregate`)
    - The aggregate attestation defined by `hash_tree_root(aggregate)` has _not_ already been seen (via aggregate gossip, within a block, or through the creation of an equivalent aggregate locally).
    - The block being voted for (`aggregate.data.beacon_block_root`) passes validation.
    - `aggregate.data.slot` is within the last `ATTESTATION_PROPAGATION_SLOT_RANGE` slots (with a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` allowance) -- i.e. `aggregate.data.slot + ATTESTATION_PROPAGATION_SLOT_RANGE >= current_slot >= aggregate.data.slot`.

>     - The validator index is within the aggregate's committee -- i.e. `aggregate_and_proof.aggregator_index in get_attesting_indices(state, aggregate.data, aggregate.aggregation_bits)`.

Reviewer: ```suggestion
    - The aggregator's validator index is within the aggregate's committee -- i.e. `aggregate_and_proof.aggregator_index in get_attesting_indices(state, aggregate.data, aggregate.aggregation_bits)`.
```

    - `aggregate_and_proof.selection_proof` selects the validator as an aggregator for the slot -- i.e. `is_aggregator(state, aggregate.data.slot, aggregate.data.index, aggregate_and_proof.selection_proof)` returns `True`.

>     - The `signed_aggregate_and_proof` is the first aggregate received for the aggregator for the slot, `aggregate.data.slot`

Reviewer: https://github.com/ethereum/eth2.0-specs/pull/1615/files#r377594638
```suggestion
    - The `aggregate` is the first aggregate received for the aggregator with index `aggregate_and_proof.aggregator_index` for the slot `aggregate.data.slot`.
```

>     - The `aggregate_and_proof.selection_proof` is a valid signature of the `aggregate.data.slot` by the validator with index `aggregate_and_proof.aggregator_index`.

Reviewer: Should we drop the `beacon_aggregate_and_proof` by the same aggregator at the slot? i.e.,
```suggestion
    - The `aggregate_and_proof.selection_proof` is a valid signature of the `aggregate.data.slot` by the validator with index `aggregate_and_proof.aggregator_index` that has _not_ already been seen.
```

Also, it might be more efficient to move these new aggregator signature validation steps to the first.

Reviewer 2: Isn't it addressed by the previous check?

Reviewer: @mkalinin Ah sorry, you're right, I missed "first" at the previous check. 😅

To double confirm the meaning "The `signed_aggregate_and_proof` is the first aggregate received for the aggregator for the slot, `aggregate.data.slot`": does that mean (i) "the first aggregate received for the validator with index `aggregate_and_proof.aggregator_index` for the slot", or "the first aggregate received for the aggregators for the slot for the subnet"? 

It could be modified to this, less confusing:
```
    - The `aggregate` is the first aggregate received for the aggregator with index `aggregate_and_proof.aggregator_index` for the slot `aggregate.data.slot`.
```

Author: yup! will clarify

- `committee_index{subnet_id}_beacon_attestation` - These topics are used to propagate unaggregated attestations to the subnet `subnet_id` (typically beacon and persistent committees) to be aggregated before being gossiped to `beacon_aggregate_and_proof`. The following validations MUST pass before forwarding the `attestation` on the subnet.
    - The attestation's committee index (`attestation.data.index`) is for the correct subnet.

>     - The attestation is the first attestation received for the participating validator for the slot, `attestation.data.slot`

Reviewer 3: What if there are multiple bits set in the attestation? (Or is this ruled out for now?)

Author: got the order messed up. there's a condition below that ensures that just one attester is included.
will fix

Author: done

    - `attestation.data.slot` is within the last `ATTESTATION_PROPAGATION_SLOT_RANGE` slots (within a `MAXIMUM_GOSSIP_CLOCK_DISPARITY` allowance) -- i.e. `attestation.data.slot + ATTESTATION_PROPAGATION_SLOT_RANGE >= current_slot >= attestation.data.slot`.
    - The attestation is unaggregated -- that is, it has exactly one participating validator (`len([bit for bit in attestation.aggregation_bits if bit == 0b1]) == 1`).

>     - This is the first attestation for the participating validator at this slot.

Reviewer 3: double?

Author: yep, whoops
